### PR TITLE
[feat]: [CI-12621]: Parse Buildx Logs to get cache metrics (using channels)

### PR DESCRIPTION
**Pull request by anurag-harness**
We need to parse the output of the docker buildx command and calculate cache metrics using that. In this PR we implement a Tee which takes a writer to write to and also creates another buffered channel to write to. The write function implemented by our Tee will write to stdout and to the channel. 
The ParseCacheMetrics will read from the channel and parse metrics which it then writes to a file. We will display stats for all layers as well as the status and time for each layer

Screenshot 2024-05-24 at 3 10 31 PM

### CODE REVIEW

File: docker.go [hunk 1]
```diff
@@ -119,8 +124,51 @@ type (
...
+		Layers      map[int]LayerStatus `json:"layers"`
+	}
+
+	tee struct {
+		w      io.Writer
+		status chan string
+	}
 )
 
+func (t *tee) Write(p []byte) (n int, err error) {
```
Comment (rutvijmehta-harness) [inline docker.go:148]: Add comments in this method to understand the data flow. First one going to stdout and second one going to the channel

File: docker.go [hunk 2]
```diff
@@ -119,8 +124,51 @@ type (
...
+	CacheMetrics struct {
+		TotalLayers int                 `json:"total_layers"`
+		Done        int                 `json:"done"`
+		Cached      int                 `json:"cached"`
+		Errored     int                 `json:"errored"`
+		Cancelled   int                 `json:"cancelled"`
+		Layers      map[int]LayerStatus `json:"layers"`
+	}
+
+	tee struct {
```
Comment (rutvijmehta-harness) [inline docker.go:142]: Move everything tee and its methods in a different file

File: docker.go [hunk 3]
```diff
@@ -232,8 +280,42 @@ func (p Plugin) Exec() error {
...
+				fmt.Printf("Could not parse cache metrics: %s", err)
+			}
+			wg.Wait()
 
-		err := cmd.Run()
+			if goroutineErr != nil {
+				return goroutineErr
+			}
+
+			if err := writeCacheMetrics(cacheMetrics, p.CacheMetricsFile); err != nil {
```
Comment (rutvijmehta-harness) [inline docker.go:313]: You should only write when you're able to pass successfully

File: docker.go [hunk 4]
```diff
@@ -281,6 +363,64 @@ func (p Plugin) Exec() error {
...
+func parseCacheMetrics(ch <-chan string) (CacheMetrics, error) {
+	var cacheMetrics CacheMetrics
+	cacheMetrics.Layers = make(map[int]LayerStatus) // Initialize the map
+
+	re := regexp.MustCompile(`#(\d+) (DONE|CACHED|ERRORED|CANCELLED)(?: ([0-9.]+)s)?`)
+
+	for line := range ch {
+		matches := re.FindAllStringSubmatch(line, -1)
+		for _, match := range matches {
+			if len(match) > 2 {
```
Comment (rutvijmehta-harness) [inline docker.go:375]: Add an early continue to save on indentation

File: docker.go [hunk 5]
```diff
@@ -281,6 +363,64 @@ func (p Plugin) Exec() error {
...
+				layerIndex, _ := strconv.Atoi(match[1])
```
Comment (rutvijmehta-harness) [inline docker.go:376]: Ignoring the error?

File: docker.go (hunk 1 above)
Comment (anurag-harness) [reply docker.go]: Done

File: docker.go (hunk 2 above)
Comment (anurag-harness) [reply docker.go]: Done

File: docker.go (hunk 3 above)
Comment (anurag-harness) [reply docker.go]: Done

File: docker.go (hunk 5 above)
Comment (anurag-harness) [reply docker.go]: Not anymore

File: docker.go (hunk 4 above)
Comment (anurag-harness) [reply docker.go]: Done

**Review by rutvijmehta-harness — APPROVED** [review]

**Pull request merged**